### ISSUE TITLE
“expected SetHash but got Hash” in Jupyter::Kernel (+ Crane)

### ISSUE THREAD

**Issue by AlexDaniel** [rakudo/rakudo#2721]
[Jupyter::Kernel](https://github.com/bduggan/p6-jupyter-kernel).

Bisected: 616b757a461faa90d889b77d53e093e70280f059

Obligatory ping: @lizmat 

Output:
```
===> Searching for: Jupyter::Kernel
===> Found: Jupyter::Kernel:ver<0.0.9> [via Zef::Repository::Ecosystems<p6c>]
===> Fetching [OK]: Jupyter::Kernel:ver<0.0.9> to /home/alex/Blin/data/zef-data/tmp/1551029720.1243.1822/p6-jupyter-kernel.git
===> Extraction [OK]: Jupyter::Kernel to /home/alex/Blin/data/zef-data/store/p6-jupyter-kernel.git
===> Testing: Jupyter::Kernel:ver<0.0.9>
t/01-basic.t ......... ok
# Failed test 'store eval in Out[1]'
# at t/02-sandbox.t line 65
# expected: '123'
#      got: 'Type check failed in assignment to $!lexicals; expected SetHash but got Hash (${})'
# Failed test 'get Out[1]'
# at t/02-sandbox.t line 66
# expected: '123'
#      got: 'Type check failed in assignment to $!lexicals; expected SetHash but got Hash (${})'
# Failed test 'get _2'
# at t/02-sandbox.t line 67
# expected: '123'
#      got: 'Type check failed in assignment to $!lexicals; expected SetHash but got Hash (${})'
# Failed test 'get _'
# at t/02-sandbox.t line 68
# expected: '123'
#      got: 'Type check failed in assignment to $!lexicals; expected SetHash but got Hash (${})'
# Failed test 'two statements'
# at t/02-sandbox.t line 71
# expected: '100'
#      got: 'Type check failed in assignment to $!lexicals; expected SetHash but got Hash (${})'
# Failed test 'saved the right thing'
# at t/02-sandbox.t line 72
# expected: '100'
#      got: '(Any)'
# Failed test 'used _ in an expression'
# at t/02-sandbox.t line 73
# expected: '101'
#      got: '1'
# Failed test 'got right output when ending with a comment'
# at t/02-sandbox.t line 81
# expected: 'hi'
#      got: 'Type check failed in assignment to $!lexicals; expected SetHash but got Hash (${})'
# Failed test 'No output for multiple say's'
# at t/02-sandbox.t line 91
# Looks like you failed 9 tests of 50
t/02-sandbox.t ....... 
Dubious, test returned 9 (wstat 2304, 0x900)
Failed 9/50 subtests 
t/03-service.t ....... ok
t/04-completions.t ... skipped: Set P6_JUPYTER_TEST_AUTOCOMPLETE to run these
t/05-autocomplete.t .. ok
t/06-magic.t ......... ok
t/07-comms.t ......... ok
t/99-meta.t .......... ok

Test Summary Report
-------------------
t/02-sandbox.t     (Wstat: 2304 Tests: 50 Failed: 9)
  Failed tests:  29-32, 34-36, 41, 47
  Non-zero exit status: 9
Files=8, Tests=145, 64 wallclock secs ( 0.07 usr  0.01 sys + 55.48 cusr  1.81 csys = 57.37 CPU)
Result: FAIL
===> Testing [FAIL]: Jupyter::Kernel:ver<0.0.9>
Failed to get passing tests, but continuing with --force-test
===> Installing: Jupyter::Kernel:ver<0.0.9>
===> Install [OK] for Jupyter::Kernel:ver<0.0.9>

1 bin/ script [jupyter-kernel.p6] installed to:
/home/alex/Blin/installed/Jupyter::Kernel_0.0.9/bin
```

**Comment by AlexDaniel** [comment]
BTW I'm not entirely sure if it's a bug in Rakudo in this case, just reporting that the tests of the module started failing.

**Comment by AlexDaniel** [comment]
Ah, `Crane` is also unhappy about that commit:

Crane – Fail, Bisected: 616b757a461faa90d889b77d53e093e70280f059

```
===> Searching for: Crane
===> Found: Crane:ver<0.1.0>:auth<atweiden> [via Zef::Repository::Ecosystems<p6c>]
===> Fetching [OK]: Crane:ver<0.1.0>:auth<atweiden> to /home/alex/Blin/data/zef-data/tmp/1551027385.30792.1134/crane.git
===> Extraction [OK]: Crane to /home/alex/Blin/data/zef-data/store/crane.git
===> Testing: Crane:ver<0.1.0>:auth<atweiden>
✗ Crane error: add requested modifying an immutable Associative
  in block  at /home/alex/Blin/data/zef-data/store/crane.git/d8d68719f1aaba7491cb64e897fab08ccf85f683/lib/Crane/Add.pm6 (Crane::Add) line 41
  in any  at /home/alex/Blin/data/zef-data/store/crane.git/d8d68719f1aaba7491cb64e897fab08ccf85f683/lib/Crane/Add.pm6 (Crane::Add) line 31
  in sub add-to-associative at /home/alex/Blin/data/zef-data/store/crane.git/d8d68719f1aaba7491cb64e897fab08ccf85f683/lib/Crane/Add.pm6 (Crane::Add) line 331
  in sub add at /home/alex/Blin/data/zef-data/store/crane.git/d8d68719f1aaba7491cb64e897fab08ccf85f683/lib/Crane/Add.pm6 (Crane::Add) line 178
  in method add at /home/alex/Blin/data/zef-data/store/crane.git/d8d68719f1aaba7491cb64e897fab08ccf85f683/lib/Crane/Add.pm6 (Crane::Add) line 54
  in method add at /home/alex/Blin/data/zef-data/store/crane.git/d8d68719f1aaba7491cb64e897fab08ccf85f683/lib/Crane.pm6 (Crane) line 100
  in block <unit> at t/methods/add.t line 46

t/methods/add.t ........ 
Dubious, test returned 1 (wstat 256, 0x100)
Failed 6/7 subtests 
t/methods/at.t ......... ok
t/methods/copy.t ....... ok
t/methods/exists.t ..... ok
t/methods/flatten.t .... ok
t/methods/get.t ........ ok
t/methods/in.t ......... ok
t/methods/list.t ....... ok
t/methods/move.t ....... ok
t/methods/patch.t ...... ok
t/methods/remove.t ..... ok
t/methods/replace.t .... ok
t/methods/set.t ........ ok
t/methods/test.t ....... ok
t/methods/transform.t .. ok

Test Summary Report
-------------------
t/methods/add.t      (Wstat: 256 Tests: 1 Failed: 0)
  Non-zero exit status: 1
  Parse errors: Bad plan.  You planned 7 tests but ran 1.
Files=15, Tests=44, 71 wallclock secs ( 0.15 usr  0.02 sys + 65.10 cusr  2.94 csys = 68.21 CPU)
Result: FAIL
===> Testing [FAIL]: Crane:ver<0.1.0>:auth<atweiden>
Failed to get passing tests, but continuing with --force-test
===> Installing: Crane:ver<0.1.0>:auth<atweiden>
===> Install [OK] for Crane:ver<0.1.0>:auth<atweiden>
```

**Comment by lizmat** [comment]
Fixed with e68e38e047 by basically reverting stuff.  Not sure what would need to be tested to make sure this doesn't happen again, I didn't get deep enough into Crane for that.
